File: android/src/main/java/com/margelo/nitro/framegrab/FrameGrabSupport.kt
package com.margelo.nitro.framegrab

import android.content.Context
import android.net.Uri
import android.system.ErrnoException
import android.system.Os
import android.system.OsConstants
import android.system.StructStat
import java.io.File
import java.util.ArrayDeque
import java.util.concurrent.Executors
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

// region Errors

internal object FrameGrabCode {
    const val INVALID_ARGUMENT = "E_INVALID_ARGUMENT"
    const val UNSUPPORTED_URI = "E_UNSUPPORTED_URI"
    const val SOURCE_UNREADABLE = "E_SOURCE_UNREADABLE"
    const val TIMESTAMP_OUT_OF_RANGE = "E_TIMESTAMP_OUT_OF_RANGE"
    const val FRAME_EXTRACTION = "E_FRAME_EXTRACTION"
    const val ENCODE_FAILED = "E_ENCODE_FAILED"
    const val DESTINATION_WRITE = "E_DESTINATION_WRITE"
    const val DESTINATION_BUSY = "E_DESTINATION_BUSY"
    const val BUSY = "E_BUSY"
    const val INTERNAL = "E_INTERNAL"
}

internal class FrameGrabException(
    val code: String,
    message: String,
) : Exception(message)

/** Strips credentials and query strings from remote URIs. Signed URLs are secrets. */
internal fun redact(uri: String): String {
    val parsed = runCatching { Uri.parse(uri) }.getOrNull() ?: return "<unparseable uri>"
    val scheme = parsed.scheme?.lowercase()
    if (scheme != "http" && scheme != "https") return uri
    return Uri.Builder()
        .scheme(parsed.scheme)
        .authority(parsed.host ?: "<redacted>")
        .path(parsed.path)
        .build()
        .toString()
}

// endregion

// region Scheduler

/**
 * One module-wide bounded dispatcher.
 *
 * A job holds a concurrency slot *and* a destination reservation from admission
 * until completion. Admission failures leave no state behind.
 */
internal object FrameGrabDispatcher {
    /** Admission control, not cancellation: beyond this we reject with `E_BUSY`. */
    private const val MAX_PENDING = 64

    private val lock = ReentrantLock()
    private var active = 0
    private var storedMaxConcurrency = 2
    private val pending = ArrayDeque<Runnable>()
    private val reservations = HashSet<String>()

    /** Gated by [storedMaxConcurrency], so this never grows past the active bound. */
    private val workers =
        ThreadPoolExecutor(
            0,
            Int.MAX_VALUE,
            60L,
            TimeUnit.SECONDS,
            SynchronousQueue<Runnable>(),
        ) { runnable ->
            Thread(runnable, "frame-grab-worker").apply { isDaemon = true }
        }

    /**
     * Argument validation and path resolution touch the filesystem, so they must
     * not run on the JS thread either. They are cheap and serialized here so the
     * destination reservation is taken before anything is queued.
     */
    private val intake =
        Executors.newSingleThreadExecutor { runnable ->
            Thread(runnable, "frame-grab-intake").apply { isDaemon = true }
        }

    var maxConcurrency: Int
        get() = lock.withLock { storedMaxConcurrency }
        set(value) {
            val clamped = value.coerceIn(1, 16)
            val toStart = ArrayList<Runnable>()
            lock.withLock {
                storedMaxConcurrency = clamped
                // Raising the bound should let queued jobs start immediately.
                while (active < storedMaxConcurrency && pending.isNotEmpty()) {
                    active += 1
                    toStart.add(pending.removeFirst())
                }
            }
            toStart.forEach { workers.execute(it) }
        }

    fun onIntake(block: () -> Unit) {
        intake.execute { block() }
    }

    /**
     * Reserves [key] and schedules [body], which receives the slot wait in
     * milliseconds. Throws [FrameGrabException] if the request is not admitted.
     */
    fun submit(
        key: String,
        body: (queueWaitMs: Double) -> Unit,
    ) {
        val queuedAt = System.nanoTime()
        val task =
            Runnable {
                try {
                    body((System.nanoTime() - queuedAt) / 1_000_000.0)
                } finally {
                    finish(key)
                }
            }

        lock.withLock {
            if (!reservations.add(key)) {
                throw FrameGrabException(
                    FrameGrabCode.DESTINATION_BUSY,
                    "Another extraction is already pending or running for this destination.",
                )
            }
            when {
                active < storedMaxConcurrency -> active += 1
                pending.size >= MAX_PENDING -> {
                    reservations.remove(key)
                    throw FrameGrabException(
                        FrameGrabCode.BUSY,
                        "Too many queued extractions (limit $MAX_PENDING).",
                    )
                }
                else -> {
                    pending.addLast(task)
                    return
                }
            }
        }
        workers.execute(task)
    }

    /** Releases the reservation and hands the slot to the next queued job. */
    private fun finish(key: String) {
        val next =
            lock.withLock {
                reservations.remove(key)
                val queued = pending.pollFirst()
                if (queued == null) active -= 1
                queued
            }
        next?.let { workers.execute(it) }
    }
}

// endregion

// region URI resolution

internal sealed class FrameGrabSource {
    data class LocalFile(val file: File) : FrameGrabSource()

    data class Content(val uri: Uri) : FrameGrabSource()

    data class Remote(val uri: String) : FrameGrabSource()
}

internal object FrameGrabPaths {
    fun resolveSource(raw: String): FrameGrabSource {
        if (raw.startsWith("/")) return FrameGrabSource.LocalFile(File(raw))

        val uri =
            runCatching { Uri.parse(raw) }.getOrNull()
                ?: throw FrameGrabException(
                    FrameGrabCode.UNSUPPORTED_URI,
                    "`sourceUri` could not be parsed as a URI.",
                )

        return when (uri.scheme?.lowercase()) {
            "file" -> {
                // `Uri.getPath()` returns the decoded path, so percent-encoded
                // filenames and spaces survive the round trip intact.
                val path =
                    uri.path
                        ?: throw FrameGrabException(
                            FrameGrabCode.UNSUPPORTED_URI,
                            "Malformed file URI.",
                        )
                FrameGrabSource.LocalFile(File(path))
            }
            "content" -> FrameGrabSource.Content(uri)
            "http", "https" -> FrameGrabSource.Remote(raw)
            null ->
                throw FrameGrabException(
                    FrameGrabCode.UNSUPPORTED_URI,
                    "`sourceUri` must be an absolute path or an absolute URI.",
                )
            else ->
                throw FrameGrabException(
                    FrameGrabCode.UNSUPPORTED_URI,
                    "Unsupported source scheme \"${uri.scheme}\".",
                )
        }
    }

    fun resolveDestination(raw: String): File {
        if (raw.startsWith("/")) return File(raw)

        val uri =
            runCatching { Uri.parse(raw) }.getOrNull()
                ?: throw FrameGrabException(
                    FrameGrabCode.UNSUPPORTED_URI,
                    "`destinationUri` could not be parsed as a URI.",
                )

        val scheme = uri.scheme?.lowercase()
        if (scheme != "file") {
            throw FrameGrabException(
                FrameGrabCode.UNSUPPORTED_URI,
                "`destinationUri` must be a local file; got scheme \"${uri.scheme ?: "none"}\".",
            )
        }
        return File(
            uri.path
                ?: throw FrameGrabException(
                    FrameGrabCode.UNSUPPORTED_URI,
                    "Malformed destination file URI.",
                ),
        )
    }

    /**
     * The parent directory must already exist; we never create directory trees.
     * A directory or symlink destination is rejected so we cannot be redirected
     * out of the caller's own folder.
     */
    fun validateDestination(destination: File) {
        val parent =
            destination.absoluteFile.parentFile
                ?: throw FrameGrabException(
                    FrameGrabCode.DESTINATION_WRITE,
                    "`destinationUri` has no parent directory.",
                )
        if (!parent.isDirectory) {
            throw FrameGrabException(
                FrameGrabCode.DESTINATION_WRITE,
                "Destination parent directory does not exist: ${parent.path}",
            )
        }

        val stat = lstatOrNull(destination.absolutePath) ?: return
        if (OsConstants.S_ISLNK(stat.st_mode)) {
            throw FrameGrabException(
                FrameGrabCode.DESTINATION_WRITE,
                "Destination is a symbolic link.",
            )
        }
        if (!OsConstants.S_ISREG(stat.st_mode)) {
            throw FrameGrabException(
                FrameGrabCode.DESTINATION_WRITE,
                "Destination exists and is not a regular file.",
            )
        }
    }

    /**
     * Never overwrite the input video.
     *
     * Content providers can hide arbitrary aliases behind a URI; we identify what
     * a descriptor lets us identify and document the rest.
     */
    fun assertDistinct(
        source: FrameGrabSource,
        destination: File,
        context: Context?,
    ) {
        val destinationStat = statOrNull(destination.absolutePath) ?: return

        when (source) {
            is FrameGrabSource.LocalFile -> {
                val samePath =
                    runCatching {
                        source.file.canonicalPath == destination.canonicalPath
                    }.getOrDefault(false)
                val sourceStat = statOrNull(source.file.absolutePath)
                val sameNode =
                    sourceStat != null &&
                        sourceStat.st_dev == destinationStat.st_dev &&
                        sourceStat.st_ino == destinationStat.st_ino
                if (samePath || sameNode) {
                    throw FrameGrabException(
                        FrameGrabCode.INVALID_ARGUMENT,
                        "`sourceUri` and `destinationUri` resolve to the same file.",
                    )
                }
            }
            is FrameGrabSource.Content -> {
                var sameNode = false
                runCatching {
                    context?.contentResolver?.openFileDescriptor(source.uri, "r")?.use { descriptor ->
                        val sourceStat = Os.fstat(descriptor.fileDescriptor)
                        sameNode =
                            sourceStat.st_dev == destinationStat.st_dev &&
                            sourceStat.st_ino == destinationStat.st_ino
                    }
                }
                if (sameNode) {
                    throw FrameGrabException(
                        FrameGrabCode.INVALID_ARGUMENT,
                        "`sourceUri` and `destinationUri` resolve to the same file.",
                    )
                }
            }
            is FrameGrabSource.Remote -> Unit
        }
    }

    /** Canonical reservation key: resolved parent directory + destination filename. */
    fun reservationKey(destination: File): String {
        val absolute = destination.absoluteFile
        val parent =
            runCatching { absolute.parentFile?.canonicalPath }.getOrNull()
                ?: absolute.parent
                ?: ""
        return "$parent/${absolute.name}"
    }

    private fun statOrNull(path: String): StructStat? =
        try {
            Os.stat(path)
        } catch (_: ErrnoException) {
            null
        }

    private fun lstatOrNull(path: String): StructStat? =
        try {
            Os.lstat(path)
        } catch (_: ErrnoException) {
            null
        }
}

// endregion

File: ios/FrameGrabSupport.swift
//
//  FrameGrabSupport.swift
//  NitroFrameGrab
//
//  Error contract, module-wide scheduler, and URI/destination resolution.
//

import Foundation

// MARK: - Errors

enum FrameGrabCode: String {
  case invalidArgument = "E_INVALID_ARGUMENT"
  case unsupportedUri = "E_UNSUPPORTED_URI"
  case sourceUnreadable = "E_SOURCE_UNREADABLE"
  case timestampOutOfRange = "E_TIMESTAMP_OUT_OF_RANGE"
  case frameExtraction = "E_FRAME_EXTRACTION"
  case encodeFailed = "E_ENCODE_FAILED"
  case destinationWrite = "E_DESTINATION_WRITE"
  case destinationBusy = "E_DESTINATION_BUSY"
  case busy = "E_BUSY"
  case internalError = "E_INTERNAL"
}

struct FrameGrabFailure: Error {
  let code: FrameGrabCode
  let message: String

  init(_ code: FrameGrabCode, _ message: String) {
    self.code = code
    self.message = message
  }
}

/// Keep native error identities when bridging to JS. AVFoundation often hides
/// the useful OSStatus several levels down. Do not dump underlying userInfo:
/// it can contain signed URLs and other sensitive source details.
func frameGrabDescribe(_ error: Error) -> String {
  let ns = error as NSError
  var parts = ["\(ns.localizedDescription) [\(ns.domain) \(ns.code)]"]
  var seen: Set<ObjectIdentifier> = [ObjectIdentifier(ns)]
  var current = ns
  while let underlying = current.userInfo[NSUnderlyingErrorKey] as? NSError,
    seen.insert(ObjectIdentifier(underlying)).inserted
  {
    parts.append("underlying \(underlying.domain) \(underlying.code)")
    current = underlying
  }
  return parts.joined(separator: " — ")
}

/// Strips credentials and query strings from remote URIs before they reach a
/// log line or a JS error message. Signed URLs are secrets.
func frameGrabRedact(_ uri: String) -> String {
  guard var components = URLComponents(string: uri),
    let scheme = components.scheme?.lowercased(),
    scheme == "http" || scheme == "https"
  else {
    return uri
  }
  components.user = nil
  components.password = nil
  components.query = nil
  components.fragment = nil
  return components.string ?? "\(scheme)://<redacted>"
}

// MARK: - Scheduler

/// One module-wide bounded dispatcher.
///
/// Holds two things for the lifetime of a job:
/// - a concurrency slot (`maxConcurrency`, default 2)
/// - a destination reservation, keyed by the canonical destination path
///
/// Admission failures never leave state behind; `release` is the only exit.
final class FrameGrabScheduler: @unchecked Sendable {
  static let shared = FrameGrabScheduler()

  /// Admission control, not cancellation: beyond this we reject with `E_BUSY`.
  static let maxPending = 64

  private struct Waiter {
    let continuation: CheckedContinuation<Void, Error>
  }

  private let lock = NSLock()
  private var active = 0
  private var storedMaxConcurrency = 2
  private var pending: [Waiter] = []
  private var reservations = Set<String>()

  private init() {}

  var maxConcurrency: Int {
    get {
      lock.lock()
      defer { lock.unlock() }
      return storedMaxConcurrency
    }
    set {
      let clamped = max(1, min(16, newValue))
      var toResume: [Waiter] = []
      lock.lock()
      storedMaxConcurrency = clamped
      // Raising the bound should let queued jobs start immediately.
      while active < storedMaxConcurrency, !pending.isEmpty {
        active += 1
        toResume.append(pending.removeFirst())
      }
      lock.unlock()
      toResume.forEach { $0.continuation.resume() }
    }
  }

  /// Reserves `key` and waits for a slot. Returns the queue wait in milliseconds.
  /// Throws `E_DESTINATION_BUSY` or `E_BUSY` without reserving anything.
  func acquire(key: String) async throws -> Double {
    let start = DispatchTime.now()
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      lock.lock()

      guard reservations.insert(key).inserted else {
        lock.unlock()
        continuation.resume(
          throwing: FrameGrabFailure(
            .destinationBusy,
            "Another extraction is already pending or running for this destination."))
        return
      }

      if active < storedMaxConcurrency {
        active += 1
        lock.unlock()
        continuation.resume()
        return
      }

      if pending.count >= Self.maxPending {
        reservations.remove(key)
        lock.unlock()
        continuation.resume(
          throwing: FrameGrabFailure(
            .busy, "Too many queued extractions (limit \(Self.maxPending))."))
        return
      }

      pending.append(Waiter(continuation: continuation))
      lock.unlock()
    }
    return Double(DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000
  }

  /// Releases the reservation and hands the slot to the next queued job.
  func release(key: String) {
    lock.lock()
    reservations.remove(key)
    if pending.isEmpty {
      active -= 1
      lock.unlock()
      return
    }
    let next = pending.removeFirst()
    lock.unlock()
    next.continuation.resume()
  }
}

// MARK: - URI resolution

enum FrameGrabSource {
  case localFile(URL)
  case remote(URL)

  var url: URL {
    switch self {
    case .localFile(let url), .remote(let url): return url
    }
  }
}

enum FrameGrabPaths {
  /// `URL(string:)` returns nil for unencoded characters such as spaces.
  /// `.urlFragmentAllowed` preserves existing percent-escapes, so this is safe
  /// to run on already-encoded input.
  static func parse(_ raw: String) -> URL? {
    if let url = URL(string: raw) { return url }
    guard let encoded = raw.addingPercentEncoding(withAllowedCharacters: .urlFragmentAllowed)
    else { return nil }
    return URL(string: encoded)
  }

  static func resolveSource(_ raw: String) throws -> FrameGrabSource {
    if raw.hasPrefix("/") {
      return .localFile(URL(fileURLWithPath: raw).standardizedFileURL)
    }
    guard let url = parse(raw), let scheme = url.scheme?.lowercased() else {
      throw FrameGrabFailure(
        .unsupportedUri,
        "`sourceUri` must be an absolute path or an absolute URI with a supported scheme.")
    }
    switch scheme {
    case "file":
      guard url.isFileURL else {
        throw FrameGrabFailure(.unsupportedUri, "Malformed file URI.")
      }
      return .localFile(url.standardizedFileURL)
    case "http", "https":
      return .remote(url)
    case "content":
      throw FrameGrabFailure(
        .unsupportedUri, "`content://` sources are supported on Android only.")
    default:
      throw FrameGrabFailure(.unsupportedUri, "Unsupported source scheme \"\(scheme)\".")
    }
  }

  static func resolveDestination(_ raw: String) throws -> URL {
    if raw.hasPrefix("/") {
      return URL(fileURLWithPath: raw).standardizedFileURL
    }
    guard let url = parse(raw), let scheme = url.scheme?.lowercased() else {
      throw FrameGrabFailure(
        .unsupportedUri, "`destinationUri` must be an absolute path or a local `file://` URI.")
    }
    guard scheme == "file", url.isFileURL else {
      throw FrameGrabFailure(
        .unsupportedUri,
        "`destinationUri` must be a local file; got scheme \"\(scheme)\".")
    }
    return url.standardizedFileURL
  }

  /// Verifies the parent directory exists and the destination itself is either
  /// absent or a plain regular file. Directories and symlinks are rejected so
  /// we never follow a link out of the caller's own directory.
  static func validateDestination(_ destination: URL) throws {
    let fileManager = FileManager.default
    let parent = destination.deletingLastPathComponent()

    var parentIsDirectory: ObjCBool = false
    guard fileManager.fileExists(atPath: parent.path, isDirectory: &parentIsDirectory),
      parentIsDirectory.boolValue
    else {
      throw FrameGrabFailure(
        .destinationWrite,
        "Destination parent directory does not exist: \(parent.path)")
    }

    // `attributesOfItem` does not follow symlinks, which is exactly what we want.
    guard let attributes = try? fileManager.attributesOfItem(atPath: destination.path) else {
      return  // Destination does not exist yet: nothing more to check.
    }
    let type = attributes[.type] as? FileAttributeType
    guard type == .typeRegular else {
      throw FrameGrabFailure(
        .destinationWrite,
        "Destination exists and is not a regular file (\(type?.rawValue ?? "unknown")).")
    }
  }

  /// Rejects a source and destination that name the same file, including
  /// hardlinks and symlink aliases on the same volume.
  static func assertDistinct(source: FrameGrabSource, destination: URL) throws {
    guard case .localFile(let sourceURL) = source else { return }

    let resolvedSource = sourceURL.resolvingSymlinksInPath().standardizedFileURL.path
    let resolvedDestination = destination.resolvingSymlinksInPath().standardizedFileURL.path
    if resolvedSource == resolvedDestination {
      throw FrameGrabFailure(
        .invalidArgument, "`sourceUri` and `destinationUri` resolve to the same file.")
    }

    var sourceStat = stat()
    var destinationStat = stat()
    guard stat(resolvedSource, &sourceStat) == 0, stat(resolvedDestination, &destinationStat) == 0
    else {
      return  // One of them does not exist; the path comparison above is enough.
    }
    if sourceStat.st_dev == destinationStat.st_dev, sourceStat.st_ino == destinationStat.st_ino {
      throw FrameGrabFailure(
        .invalidArgument, "`sourceUri` and `destinationUri` resolve to the same file.")
    }
  }

  /// Canonical reservation key: resolved parent directory + destination filename.
  static func reservationKey(for destination: URL) -> String {
    let parent = destination.deletingLastPathComponent().resolvingSymlinksInPath()
      .standardizedFileURL.path
    return parent + "/" + destination.lastPathComponent
  }
}

/// Escape hatch for carrying non-`Sendable` CoreGraphics/Foundation values into
/// a `DispatchQueue` closure. Ownership is single-threaded by construction: the
/// value is produced, moved once, and never touched concurrently.
struct FrameGrabBox<T>: @unchecked Sendable {
  let value: T
  init(_ value: T) { self.value = value }
}
